File: rubygems/pray-cli/lib/pray/upstream_merge.rb
# frozen_string_literal: true

module Pray
  module Upstream
    module_function

    def identity_path?(path)
      File.extname(path.to_s) == ".prayspec"
    end

    def content_paths(files)
      files.reject { |path| identity_path?(path) }
    end

    def clean_replica?(old_content, local_content)
      local_content == old_content
    end

    def try_merge_content_files(old_content, new_content, local_content)
      return [new_content.dup, []] if clean_replica?(old_content, local_content)

      merged = {}
      conflicts = []
      (old_content.keys + new_content.keys + local_content.keys).uniq.sort.each do |path|
        result = merge_one_content_path(old_content[path], new_content[path], local_content[path])
        if result == :conflict
          conflicts << path
        elsif result != :omit
          merged[path] = result
        end
      end
      [merged, conflicts]
    end

    def merge_one_content_path(old, new, local)
      case [old.nil?, new.nil?, local.nil?]
      when [false, false, false]
        merge_all_present(old, new, local)
      when [false, true, false]
        (local == old) ? :omit : :conflict
      when [false, false, true]
        (old == new) ? :omit : new
      when [true, false, true]
        new
      when [true, true, false]
        local
      when [true, false, false]
        (local == new) ? new : :conflict
      else
        :conflict
      end
    end

    def merge_all_present(old, new, local)
      return new if local == new || local == old
      return local if old == new

      :conflict
    end

    def merge_content_files(old_content, new_content, local_content)
      merged, conflicts = try_merge_content_files(old_content, new_content, local_content)
      return merged if conflicts.empty?

      raise Error.resolution("upstream merge conflict in #{conflicts.join(", ")}")
    end

    def merge_conflict_message(fork, upstream_name, old_version, new_version, paths)
      "upstream merge conflict in #{fork} while refreshing #{upstream_name} " \
        "#{old_version} to #{new_version}: #{paths.join(", ")}"
    end

    def next_upstream_constraint(current, new_version)
      trimmed = current.to_s.strip
      return current if trimmed == "*" || trimmed.start_with?("~>", "^", ">=", ">", "<=", "<")

      "= #{new_version}"
    end
    private_class_method :merge_one_content_path, :merge_all_present
  end
end

File: npmjs/pray-cli/src/resolve/upstream-merge.ts
import { extname } from "node:path";
import { PrayError } from "../errors.js";

export function isIdentityPath(path: string): boolean {
  return extname(path) === ".prayspec";
}

export function contentPaths(files: readonly string[]): string[] {
  return files.filter((path) => !isIdentityPath(path));
}

export function isCleanReplica(
  oldContent: Map<string, Buffer>,
  localContent: Map<string, Buffer>,
): boolean {
  if (oldContent.size !== localContent.size) {
    return false;
  }
  for (const [path, bytes] of oldContent.entries()) {
    const local = localContent.get(path);
    if (local === undefined || !local.equals(bytes)) {
      return false;
    }
  }
  return true;
}

export function tryMergeContentFiles(
  oldContent: Map<string, Buffer>,
  newContent: Map<string, Buffer>,
  localContent: Map<string, Buffer>,
): Map<string, Buffer> | string[] {
  if (isCleanReplica(oldContent, localContent)) {
    return new Map(newContent);
  }
  const paths = new Set([
    ...oldContent.keys(),
    ...newContent.keys(),
    ...localContent.keys(),
  ]);
  const merged = new Map<string, Buffer>();
  const conflicts: string[] = [];
  for (const path of [...paths].sort()) {
    const oldBytes = oldContent.get(path);
    const newBytes = newContent.get(path);
    const localBytes = localContent.get(path);
    if (
      newBytes !== undefined &&
      localBytes !== undefined &&
      localBytes.equals(newBytes)
    ) {
      merged.set(path, newBytes);
    } else if (
      oldBytes !== undefined &&
      newBytes !== undefined &&
      localBytes !== undefined &&
      localBytes.equals(oldBytes)
    ) {
      merged.set(path, newBytes);
    } else if (
      oldBytes !== undefined &&
      newBytes !== undefined &&
      localBytes !== undefined &&
      oldBytes.equals(newBytes)
    ) {
      merged.set(path, localBytes);
    } else if (
      oldBytes !== undefined &&
      newBytes === undefined &&
      localBytes !== undefined &&
      localBytes.equals(oldBytes)
    ) {
    } else if (
      oldBytes !== undefined &&
      newBytes !== undefined &&
      localBytes === undefined &&
      oldBytes.equals(newBytes)
    ) {
    } else if (
      oldBytes === undefined &&
      newBytes !== undefined &&
      localBytes === undefined
    ) {
      merged.set(path, newBytes);
    } else if (
      oldBytes === undefined &&
      newBytes === undefined &&
      localBytes !== undefined
    ) {
      merged.set(path, localBytes);
    } else if (
      oldBytes !== undefined &&
      newBytes !== undefined &&
      localBytes === undefined
    ) {
      merged.set(path, newBytes);
    } else {
      conflicts.push(path);
    }
  }
  return conflicts.length === 0 ? merged : conflicts;
}

export function mergeContentFiles(
  oldContent: Map<string, Buffer>,
  newContent: Map<string, Buffer>,
  localContent: Map<string, Buffer>,
): Map<string, Buffer> {
  const merged = tryMergeContentFiles(oldContent, newContent, localContent);
  if (merged instanceof Map) {
    return merged;
  }
  throw PrayError.resolution(`upstream merge conflict in ${merged.join(", ")}`);
}

export function upstreamMergeConflictMessage(
  fork: string,
  upstreamName: string,
  oldVersion: string,
  newVersion: string,
  paths: readonly string[],
): string {
  return `upstream merge conflict in ${fork} while refreshing ${upstreamName} ${oldVersion} to ${newVersion}: ${paths.join(", ")}`;
}

export type OverlayFileChange = "changed" | "local_only" | "missing";

export function overlayFileChanges(
  localContent: Map<string, Buffer>,
  upstreamContent: Map<string, Buffer>,
): Array<[string, OverlayFileChange]> {
  const paths = new Set([...localContent.keys(), ...upstreamContent.keys()]);
  const changes: Array<[string, OverlayFileChange]> = [];
  for (const path of [...paths].sort()) {
    const local = localContent.get(path);
    const upstream = upstreamContent.get(path);
    if (
      local !== undefined &&
      upstream !== undefined &&
      !local.equals(upstream)
    ) {
      changes.push([path, "changed"]);
    } else if (local !== undefined && upstream === undefined) {
      changes.push([path, "local_only"]);
    } else if (local === undefined && upstream !== undefined) {
      changes.push([path, "missing"]);
    }
  }
  return changes;
}

export function overlayDriftLine(
  fork: string,
  upstreamName: string,
  upstreamVersion: string,
  path: string,
  change: OverlayFileChange,
): string {
  if (change === "changed") {
    return `${fork} ${path} differs from ${upstreamName} ${upstreamVersion}`;
  }
  if (change === "local_only") {
    return `${fork} ${path} local`;
  }
  return `${fork} ${path} missing from fork`;
}

export function nextUpstreamConstraint(
  current: string,
  newVersion: string,
): string {
  const trimmed = current.trim();
  if (
    trimmed === "*" ||
    trimmed.startsWith("~>") ||
    trimmed.startsWith("^") ||
    trimmed.startsWith(">=") ||
    trimmed.startsWith(">") ||
    trimmed.startsWith("<=") ||
    trimmed.startsWith("<")
  ) {
    return current;
  }
  return `= ${newVersion}`;
}
